### ISSUE TITLE
task6

### ISSUE THREAD

**Issue by tuchang** [tuchang/PyTorch-datawhale-201908#6]
task 6 多种优化器 理解更多神经网络优化方法
1. 了解不同优化器
2. 书写优化器代码
3. Momentum
4. 二维优化，随机梯度下降法进行优化实现
5. Ada自适应梯度调节法
6. RMSProp
7. Adam
8. PyTorch多种优化器选择

* 1 torch.optim.SGD 
* 2 torch.optim.ASGD 
* 3 torch.optim.Rprop 
* 4 torch.optim.Adagrad 
* 5 torch.optim.Adadelta 
* 6 torch.optim.RMSprop 
* 7 torch.optim.Adam(AMSGrad) 
* 8 torch.optim.Adamax 
* 9 torch.optim.SparseAdam 
* 10 torch.optim.LBFGS 

Adam 

https://www.cnblogs.com/guoyaohua/p/8542554.html
https://zhuanlan.zhihu.com/p/64885176
TensorFlow 中的优化器
https://cloud.tencent.com/developer/article/1083486
https://www.tensorflow.org/api_guides/python/train

1 了解不同优化器
梯度下降最常见的三种变形 BGD，SGD，MBGD

ASGD就是用空间换时间的一种SGD

Rprop实现Rprop优化方法(弹性反向传播)，优化方法原文《Martin Riedmiller und Heinrich Braun: Rprop - A Fast Adaptive Learning Algorithm. Proceedings of the International Symposium on Computer and Information Science VII, 1992》
该优化方法适用于full-batch，不适用于mini-batch，因而在min-batch大行其道的时代里，很少见到。

实现Adagrad优化方法(Adaptive Gradient)，Adagrad是一种自适应优化方法，是自适应的为各个参数分配不同的学习率。这个学习率的变化，会受到梯度的大小和迭代次数的影响。梯度越大，学习率越小；梯度越小，学习率越大。缺点是训练后期，学习率过小，因为Adagrad累加之前所有的梯度平方作为分母。

实现Adadelta优化方法。Adadelta是Adagrad的改进。Adadelta分母中采用距离当前时间点比较近的累计项，这可以避免在训练后期，学习率过小。

实现RMSprop优化方法（Hinton提出），RMS是均方根（root meam square）的意思。RMSprop和Adadelta一样，也是对Adagrad的一种改进。RMSprop采用均方根作为分母，可缓解Adagrad学习率下降较快的问题。并且引入均方根，可以减少摆动
详细了解可读:
http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf

实现Adam(Adaptive Moment Estimation))优化方法。Adam是一种自适应学习率的优化方法，Adam利用梯度的一阶矩估计和二阶矩估计动态的调整学习率。吴老师课上说过，Adam是结合了Momentum和RMSprop，并进行了偏差修正。

实现Adamax优化方法。Adamax是对Adam增加了一个学习率上限的概念，所以也称之为Adamax。
详细了解可阅读，Adam: A Method for Stochastic Optimization(Adam: A Method for Stochastic Optimization)(没错，就是Adam论文中提出了Adamax)。

SparseAdam 针对稀疏张量的一种简化版 Adam优化方法。

LBFGS
实现L-BFGS（Limited-memory Broyden–Fletcher–Goldfarb–Shanno）优化方法。L-BFGS属于拟牛顿算法。L-BFGS是对BFGS的改进，特点就是节省内存。

Momentum 是「運動量」的意思，此優化器為模擬物理動量的概念，在同方向的維度上學習速度會變快，方向改變的時候學習速度會變慢。
